File: c++/0024.cpp
// Link: https://leetcode.com/problems/swap-nodes-in-pairs/description/
// Runtime: 3 ms
// Memory: 7.6 MB

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution 
{
public:
    ListNode* swapPairs(ListNode* head) 
    {
        if ((head == nullptr) || (head->next == nullptr)) return head;

        ListNode* nodeBeforeList = new ListNode();
        ListNode* previousNode = nodeBeforeList;
        ListNode* currentNode = head;

        while ((currentNode != nullptr) && (currentNode->next != nullptr))
        {
            previousNode->next = currentNode->next;
            currentNode->next = currentNode->next->next;
            previousNode->next->next = currentNode;

            previousNode = currentNode;
            currentNode = currentNode->next;
        }

        return nodeBeforeList->next;
    }
};

File: python/0024.py
# Link: https://leetcode.com/problems/swap-nodes-in-pairs/description/
# Runtime: 40 ms
# Memory: 16.3 MB

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head

        node = head.next
        head.next = self.swapPairs(head.next.next)
        node.next = head

        return node
